shader/flatten.wgsl

```diff
@@ -188,96 +357,118 @@ fn flatten_cubic(cubic: CubicPoints, path_ix: u32, local_to_device: Transform, o
-    let Q_ACCURACY = ACCURACY * 0.1;
-    let REM_ACCURACY = ACCURACY - Q_ACCURACY;
-    let MAX_HYPOT2 = 432.0 * Q_ACCURACY * Q_ACCURACY;
-    let scaled_sqrt_tol = sqrt(REM_ACCURACY / scale);
-    // Fudge the subdivision count metric to account for `scale` when the subdivision is done in local
-    // coordinates.
-    var n_quads = max(u32(ceil(pow(err * (1.0 / MAX_HYPOT2), 1.0 / 6.0)) * scale), 1u);
-    n_quads = min(n_quads, MAX_QUADS);
-
-    var keep_params: array<SubdivResult, MAX_QUADS>;
-    var val = 0.0;
-    var qp0 = p0;
-    let step = 1.0 / f32(n_quads);
-    for (var i = 0u; i < n_quads; i += 1u) {
-        let t = f32(i + 1u) * step;
-        let qp2 = eval_cubic(p0, p1, p2, p3, t);
-        var qp1 = eval_cubic(p0, p1, p2, p3, t - 0.5 * step);
-        qp1 = 2.0 * qp1 - 0.5 * (qp0 + qp2);
-
-        // TODO: Estimate an accurate subdivision count for strokes
-        let params = estimate_subdiv(qp0, qp1, qp2, scaled_sqrt_tol);
-        keep_params[i] = params;
-        val += params.val;
-        qp0 = qp2;
+    // Drop zero length lines.
+    if all(p0 == p1) && all(p0 == p2) && all(p0 == p3) {
+        return;
     }
-
-    // Normal vector to calculate the start point of the offset curve.
-    var n0 = offset * cubic_start_normal(p0, p1, p2, p3);
-
-    let n = max(u32(ceil(val * (0.5 / scaled_sqrt_tol))), 1u);
-    var lp0 = p0;
-    qp0 = p0;
-    let v_step = val / f32(n);
-    var n_out = 1u;
-    var val_sum = 0.0;
-    for (var i = 0u; i < n_quads; i += 1u) {
-        let t = f32(i + 1u) * step;
-        let qp2 = eval_cubic(p0, p1, p2, p3, t);
-        var qp1 = eval_cubic(p0, p1, p2, p3, t - 0.5 * step);
-        qp1 = 2.0 * qp1 - 0.5 * (qp0 + qp2);
-        let params = keep_params[i];
-        let u0 = approx_parabola_inv_integral(params.a0);
-        let u2 = approx_parabola_inv_integral(params.a2);
-        let uscale = 1.0 / (u2 - u0);
-        var val_target = f32(n_out) * v_step;
-        while n_out == n || val_target < val_sum + params.val {
-            var lp1: vec2f;
-            var t1: f32;
-            if n_out == n {
-                lp1 = p3;
-                t1 = 1.;
-            } else {
-                let u = (val_target - val_sum) / params.val;
-                let a = mix(params.a0, params.a2, u);
-                let au = approx_parabola_inv_integral(a);
-                let t = (au - u0) * uscale;
-                t1 = t;
-                lp1 = eval_quad(qp0, qp1, qp2, t);
+    //output_line_with_transform(path_ix, t_start, t_end, transform);
+    //if true { return; }
+
+    let tol = 0.25;
+    var t0_u = 0u;
+    var dt = 1.0;
+    var last_p = p0;
+    var last_q = p1 - p0;
+    if dot(last_q, last_q) < DERIV_THRESH * DERIV_THRESH {
+        last_q = eval_cubic_and_deriv(p0, p1, p2, p3, DERIV_EPS).deriv;
+    }
+    var last_t = 0.0;
+    var lp0 = t_start;
+    loop {
+        let t0 = f32(t0_u) * dt;
+        if t0 == 1.0 {
+            break;
+        }
+        loop {
```

Comment:
Should there be a short explainer here about the derivative threshold, the iterative search to handle cusps, and the 3 cases below?

Reply:
That feels more like paper than code, but I'll add a sentence each.

Reply:
I agree. I think the paper will be the in-depth reference for all of this BUT I would like the code to eventually contain explainers and relevant citations as much as possible. I think the comments can be improved but I'm OK with the current state of things.